### PR TITLE
[BB-840] Add a role to install and set up unbound

### PR DESCRIPTION
**Pull request by lgp171188**
This will be used on haproxy-stage.

### CODE REVIEW

File: playbooks/roles/unbound/defaults/main.yml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Set the zone forwardnig configuration in the 'forward_zones' list. Each entry will require a 'name' key and
```
Comment (UmanShahzad) [inline playbooks/roles/unbound/defaults/main.yml]: Nit: add `---` above

File: playbooks/roles/unbound/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,18 @@
+---
+- name: Install unbound and resolvconf
+  apt:
+    name={{ item }}
+    state=installed
```
Comment (UmanShahzad) [inline playbooks/roles/unbound/tasks/main.yml]: Let's consistently use the dictionary format with `:`

File: playbooks/roles/unbound/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  with_items:
+    - unbound
+    - resolvconf
+
+- name: Copy the configuration for caching and forwarding
+  template:
+    src: cache-and-forward.conf.j2
+    dest: "{{ unbound_config_dir }}/cache-and-forward.conf"
```
Comment (UmanShahzad) [inline playbooks/roles/unbound/tasks/main.yml]: Is `unbound_config_dir` already created upon installation above?

File: playbooks/roles/unbound/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+- name: restart unbound
```
Comment (UmanShahzad) [inline playbooks/roles/unbound/tasks/main.yml]: This should be a handler and only triggered by notification if the template changes.

File: playbooks/roles/unbound/templates/cache-and-forward.conf.j2 [hunk 5]
```diff
@@ -0,0 +1,26 @@
+#jinja2: lstrip_blocks: "True"
```
Comment (UmanShahzad) [inline playbooks/roles/unbound/templates/cache-and-forward.conf.j2]: Why is this commented out?

File: playbooks/roles/unbound/defaults/main.yml (hunk 1 above)
Comment (UmanShahzad) [inline playbooks/roles/unbound/defaults/main.yml]: Also: typo in `forwardnig`

File: playbooks/roles/unbound/defaults/main.yml [hunk 6]
```diff
@@ -0,0 +1,4 @@
...
+# a 'forward-addr'/'forward-host' key with appropriate values
```
Comment (UmanShahzad) [inline playbooks/roles/unbound/defaults/main.yml]: This comment looks like it should be above` forward_zones`, not `unbound_config_dir`? And then we can also add some spacing.

File: playbooks/roles/unbound/defaults/main.yml [hunk 7]
```diff
@@ -0,0 +1,4 @@
...
+unbound_config_dir: /etc/unbound/unbound.conf.d
+forward_zones: []
```
Comment (UmanShahzad) [inline playbooks/roles/unbound/defaults/main.yml]: A proper commented example for this would be nice.

File: playbooks/roles/unbound/templates/cache-and-forward.conf.j2 [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+server:
+    cache-max-negative-ttl: 600
```
Comment (UmanShahzad) [inline playbooks/roles/unbound/templates/cache-and-forward.conf.j2]: Can we make this configurable?

Comment (UmanShahzad) [reply playbooks/roles/unbound/templates/cache-and-forward.conf.j2]: What unit is this talking about? Can we put a comment for that?

File: playbooks/roles/unbound/templates/cache-and-forward.conf.j2 [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+forward-zone:
+    name: {{ zone.name }}
+    {% if zone.forward_addr is defined %}
+    {% if zone.forward_addr is string %}
+    forward-addr: {{ zone.forward_addr }}
+    {% else %}
+    {% for addr in zone.forward_addr %}
+    forward-addr: {{ addr }}
+    {% endfor %}
+    {% endif %}
```
Comment (UmanShahzad) [inline playbooks/roles/unbound/templates/cache-and-forward.conf.j2]: This is hard to follow without some comments - can you put a tiny comment like `{% endif %} # if zone.forward_addr is string` on the `endif` here, and on others to make it readable?

File: playbooks/roles/unbound/templates/cache-and-forward.conf.j2 [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+    {% elif zone.forward_host is defined %}
+    {% if zone.forward_host is string %}
+    forward-host: {{ zone.forward_host }}
+    {% else %}
+    {% for host in zone.forward_host %}
+    forward-host: {{ host }}
+    {% endfor %}
+    {% endif %}
+    {% endif %}
+
```
Comment (UmanShahzad) [inline playbooks/roles/unbound/templates/cache-and-forward.conf.j2]: Is this newline necessary?

File: playbooks/roles/unbound/tasks/main.yml (hunk 3 above)
Comment (lgp171188) [reply playbooks/roles/unbound/tasks/main.yml]: @UmanShahzad, yes it is.

```
# dpkg -L unbound
/.
/etc
/etc/apparmor.d
/etc/apparmor.d/usr.sbin.unbound
/etc/default
/etc/init.d
/etc/init.d/unbound
/etc/insserv.conf.d
/etc/insserv.conf.d/unbound
/etc/resolvconf
/etc/resolvconf/update.d
/etc/resolvconf/update.d/unbound
/etc/unbound
/etc/unbound/unbound.conf
/etc/unbound/unbound.conf.d
/etc/unbound/unbound.conf.d/qname-minimisation.conf
/etc/unbound/unbound.conf.d/root-auto-trust-anchor-file.conf
/lib
/lib/systemd
/lib/systemd/system
/lib/systemd/system/unbound-resolvconf.service
/lib/systemd/system/unbound.service
/usr
/usr/lib
/usr/lib/resolvconf
/usr/lib/resolvconf/dpkg-event.d
/usr/lib/resolvconf/dpkg-event.d/unbound
/usr/lib/unbound
/usr/lib/unbound/package-helper
/usr/sbin
/usr/sbin/unbound
/usr/sbin/unbound-checkconf
/usr/sbin/unbound-control
/usr/sbin/unbound-control-setup
/usr/share
/usr/share/doc
/usr/share/doc/unbound
/usr/share/doc/unbound/CREDITS
/usr/share/doc/unbound/FEATURES
/usr/share/doc/unbound/NEWS.Debian.gz
/usr/share/doc/unbound/README.DNS64
/usr/share/doc/unbound/README.gz
/usr/share/doc/unbound/TODO.gz
/usr/share/doc/unbound/changelog.Debian.gz
/usr/share/doc/unbound/changelog.gz
/usr/share/doc/unbound/contrib
/usr/share/doc/unbound/contrib/update-anchor.sh.gz
/usr/share/doc/unbound/copyright
/usr/share/doc/unbound/examples
/usr/share/doc/unbound/examples/unbound.conf
/usr/share/man
/usr/share/man/man5
/usr/share/man/man5/unbound.conf.5.gz
/usr/share/man/man8
/usr/share/man/man8/unbound-checkconf.8.gz
/usr/share/man/man8/unbound-control.8.gz
/usr/share/man/man8/unbound.8.gz
/usr/share/munin
/usr/share/munin/plugins
/usr/share/munin/plugins/unbound_munin_
/var
/var/lib
/var/lib/unbound
```

File: playbooks/roles/unbound/tasks/main.yml (hunk 4 above)
Comment (lgp171188) [reply playbooks/roles/unbound/tasks/main.yml]: @UmanShahzad, that is correct. Will fix this.

File: playbooks/roles/unbound/templates/cache-and-forward.conf.j2 (hunk 5 above)
Comment (lgp171188) [reply playbooks/roles/unbound/templates/cache-and-forward.conf.j2]: @UmanShahzad, this is the way to specify the jinja2 configuration directive.

File: playbooks/roles/unbound/templates/cache-and-forward.conf.j2 (hunk 10 above)
Comment (lgp171188) [reply playbooks/roles/unbound/templates/cache-and-forward.conf.j2]: @UmanShahzad, this newline was necessary to generate a good looking output file configuration.

**Comment by UmanShahzad** [comment]
Code looks good to me now @lgp171188 .
